Dropzone TLC

KevinTriplett opened the pull request:
closes #415 

Work so far:

- Pressing ESC cancels the drag and puts the widget back

KevinTriplett commented:
I'm thinking to experiment with the "Sortable" feature: https://docs.dndkit.com/presets/sortable

It depends on whether or not we can still drop widgets outside of containers like we currently do. I think we can and it's more what a user is expecting inside containers, imo.

tibetsprague commented:
nice. I do plan to add a delete button to each widget when in edit mode which will also be more obvious than dragging it off

KevinTriplett commented:
> nice. I do plan to add a delete button to each widget when in edit mode which will also be more obvious than dragging it off

Kewl! Yeah, that would be a great improvement

KevinTriplett commented:
I think this "disallow container drop on container" may have fixed your wild dropzones, although there's one missing dropzone at the bottom of the menu, I'm going to add it now.

thomasgwatson commented on apps/web/src/routes/AuthLayoutRouter/components/ContextMenu/ContextMenu.jsx:
```diff
@@ -461,10 +470,21 @@ function GrabMe ({ children, ...props }) {
+  const { contextWidgets, activeWidget } = useContextMenuContext()
+  const activeWidgetIsContainer = ['container', 'home', 'chats', 'members', 'setup'].includes(activeWidget?.type)
+    || activeWidget?.childWidgets?.length > 0
+  const isIllegalDrop = droppableParams.id.includes('bottom-of-child-list') && activeWidgetIsContainer
```
This is likely duplicating logic in `isValidChildWidget`.

Given the product and implementation context of the drag and drop stuff, I think it might be best for me to dig into this issue. I would have assigned myself earlier in the week but we've been deep in rare in-person time.

KevinTriplett replied:
Yeah, it's a bit complex, this seems to work but let's take a look at it and see if we can avoid dups

thomasgwatson commented:
@KevinTriplett  I've added a few more specific issues I've spotted to the original ticket. Do you want to pair on this at some point in the next day or two?

KevinTriplett commented:
> @KevinTriplett I've added a few more specific issues I've spotted to the original ticket. Do you want to pair on this at some point in the next day or two?

Sure! Here's my calendly link, it'll help us figure out time zones: https://calendly.com/kevintriplett/90-minute-project-meeting

KevinTriplett commented:
I updated the issue as well -- I see the isValidChild function. Let's see if we can combine the two, there are things I added that are not in the function. It would be good if they were all in the function. There are some things we can't do because the library doesn't give us the keyboard events and Tibet wants to remove widgets in a different way, which I think would be great.

KevinTriplett commented:
> Ahh sad about the escape keystroke inaccessibility :( I booked a time for this afternoon for us to do a session, let me know if that works for you :)

I know, right? Looking forward to working with you in a couple hours.

KevinTriplett commented:
Oh, and I pushed a commit that you can pull and try out, have a look at. There are some questions and one puzzle, in addition to hitting the rest of the issue list.

KevinTriplett commented:
@tibetsprague pls review -- one last issue that @thomasgwatson will take: although there's a dropzone at the bottom of a child list, the backend is not adding it to the end of the parent. I'll create a separate issue for Tom to take.

thomasgwatson commented on apps/web/src/routes/AuthLayoutRouter/components/ContextMenu/ContextMenu.jsx:
```diff
@@ -497,37 +516,38 @@ function DropZone ({ droppableParams, isDroppable = true, height = '', hide = fa
   }
 
-  const isActive = item.viewUser?.lastActiveAt ? new Date(parseInt(item.viewUser.lastActiveAt)) > new Date(Date.now() - 1000 * 60 * 4) : false
+  // XXX why is this using parseInt? convert to luxon?
```
Were you satisfied with my explanation Kevin?

thomasgwatson reviewed (approved):
lgtm

tibetsprague commented on packages/presenters/ContextWidgetPresenter.js:
```diff
@@ -150,23 +150,59 @@ export function getStaticMenuWidgets ({ isPublicContext, isMyContext, profileUrl
+  const parentWidgetIsContainer = containerTypes.includes(parentWidget?.type)
+  const activeWidgetIsContainer = containerTypes.includes(activeWidget?.type)
+  
+  // const listBottom = droppableParams.id?.includes('bottom-of-child-list')
```
remove?

tibetsprague reviewed (approved):
havent tested yet but code seems fine
